src/puppetlabs/trapperkeeper/bootstrap.clj

```diff
@@ -132,30 +135,168 @@
+      (catch URISyntaxException e
+        (throw (wrap-uri-error config-path e)))
+      ; Thrown by reader URI.toURL() when URI is not absolute
+      (catch IllegalArgumentException e
+        (throw (wrap-uri-error config-path e)))
+      ; Thrown if a valid URI points to a file that does not exist
+      (catch FileNotFoundException e
+        (throw (wrap-uri-error config-path e))))))
+
+(schema/defn get-annotated-bootstrap-entries :- [AnnotatedBootstrapEntry]
+  "Reads each bootstrap entry into a map with the line number and file the
+  entry is from. Returns a list of maps."
+  [configs :- [schema/Str]]
+  (for [config configs
+        [line-number line-text] (indexed (map remove-comments (read-config config)))
+        :when (not (empty? line-text))]
+    {:bootstrap-file config
+     :line-number (inc line-number)
+     :entry line-text}))
+
+(defn find-duplicates
+  "Collects duplicates base on running f on each item in coll.
+   Returns a map where the keys will be the result of running f on each item,
+   and the values will be lists of items that are duplicates of eachother"
+  [coll f]
+  (->> coll
+       (group-by f)
+       ; filter out map values with only 1 item
+       (remove #(= 1 (count (val %))))))
+
+(schema/defn duplicate-protocol-error :- IllegalArgumentException
+  "Returns an IllegalArgumentException describing what services implement
+   the same protocol, including the line number and file the bootstrap entries
+   were found"
+  [protocol-id :- schema/Keyword
+   duplicate-services :- [(schema/protocol services/ServiceDefinition)]
+   service->entry-map :- {(schema/protocol services/ServiceDefinition) AnnotatedBootstrapEntry}]
+  (let [make-error-message (fn [service]
+                             (let [entry (get service->entry-map service)]
+                               (format "%s:%s\n%s"
+                                       (:bootstrap-file entry)
+                                       (:line-number entry)
+                                       (:entry entry))))]
+    (IllegalArgumentException.
+     (format (str "Duplicate implementations found for service protocol '%s':\n%s")
+             protocol-id
+             (string/join "\n" (map make-error-message duplicate-services))))))
+
+(schema/defn check-duplicate-service-implementations!
+  "Throws an exception if two services implement the same service protocol"
+  [services :- [(schema/protocol services/ServiceDefinition)]
+   bootstrap-entries :- [AnnotatedBootstrapEntry]]
+
+  ; Zip up the services and bootstrap entries and construct a map out of them
+  ; to use as a lookup table below
+  (let [service->entry-map (zipmap services bootstrap-entries)]
+    ; Find duplicates base on the service id returned by calling service-def-id
+    ; on each service
+    (if-let [duplicate (first (find-duplicates services services/service-def-id))]
+      (throw (duplicate-protocol-error (key duplicate) (val duplicate) service->entry-map)))))
```

Comment:
So if we had duplicates for more than one protocol, would we only display the error for the first protocol where we found a duplicate?  That's probably fine - just seems like it might be nice to display all of the duplicates since `find-duplicates` would appear to actually do the work to locate them all.